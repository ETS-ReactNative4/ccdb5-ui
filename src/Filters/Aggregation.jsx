import React from 'react';
import { connect } from 'react-redux';
import CollapsibleFilter from './CollapsibleFilter';
import AggregationItem from './AggregationItem';
import './Aggregation.less';

export class Aggregation extends React.Component {
  constructor(props) {
    super(props);
    this.state = { showMore: this.props.showMore || false }
    this._toggleShowMore = this._toggleShowMore.bind(this);
  }

  _toggleShowMore() {
    this.setState({
      showMore: !this.state.showMore
    })
  }

  render() {
    const all = this.props.options || [];
    const some = all.length > 6 ? all.slice(0, 5) : all
    const remain = all.length - 6

    return (
        <CollapsibleFilter title={this.props.title}
                           desc={this.props.desc}
                           showChildren={this.props.showChildren}
                           className="aggregation">
            <ul>
            {!this.state.showMore ?
              some.map(bucket =>
                <AggregationItem item={bucket} key={ bucket.key } fieldName={ this.props.fieldName } />
            ) :
              all.map(bucket =>
                <AggregationItem item={bucket} key={ bucket.key } fieldName={ this.props.fieldName } />
            )}
            </ul>
            {remain > 0 && !this.state.showMore ? (
              <div className="flex-fixed">
                   <button className="a-btn a-btn__link hover more" onClick={ this._toggleShowMore }>+ Show {remain} more</button>
              </div>
            ) : null}
            {remain > 0 && this.state.showMore ? (
              <div className="flex-fixed">
                   <button className="a-btn a-btn__link hover more" onClick={ this._toggleShowMore }>- Show {remain} less </button>
              </div>
            ) : null}
        </CollapsibleFilter>
    );
  }
}

export const mapStateToProps = (state, ownProps) => {
  return {
    options: state.aggs[ownProps.fieldName]
  }
};

export default connect(mapStateToProps)(Aggregation);
