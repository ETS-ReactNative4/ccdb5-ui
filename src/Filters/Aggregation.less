@import (less) "../base.less";

@max-count-width: 65px;

.aggregation {
    ul {
      padding-left: 0;

      li {
        padding: 0;
        border: 0;
        margin-bottom: @gutter-narrow;

        input[type='checkbox'] {
        }

        .bucket-key {
          max-width: ~"calc(100% - 95px)";  // @max-count-width + @gutter-normal * 2
          padding-left: @gutter-normal;
        }

        .bucket-count {
          width: @max-count-width;
          text-align: right;
        }
      }
    }
}
