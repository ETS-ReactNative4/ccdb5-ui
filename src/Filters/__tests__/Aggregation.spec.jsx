import React from 'react';
import Aggregation from '../Aggregation';
import renderer from 'react-test-renderer';

describe('initial state', () => {
  it('renders without crashing', () => {
    let options = [{key: 'foo', doc_count: 99}];

    const target = renderer.create(
      <Aggregation options={options} />
    );

    let tree = target.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('show more', () => {
  let options;
  beforeEach(() => {
    options = [
      {key: 'alpha', doc_count: 99},
      {key: 'beta', doc_count: 99},
      {key: 'gamma', doc_count: 99},
      {key: 'delta', doc_count: 99},
      {key: 'epsilon', doc_count: 99},
      {key: 'zeta', doc_count: 99},
      {key: 'eta', doc_count: 99},
      {key: 'theta', doc_count: 99}
    ]
  })

  it('only shows the first 6 items of large arrays', () => {
    const target = renderer.create(
      <Aggregation options={options} />
    );

    let tree = target.toJSON();
    expect(tree).toMatchSnapshot();
  })
})

