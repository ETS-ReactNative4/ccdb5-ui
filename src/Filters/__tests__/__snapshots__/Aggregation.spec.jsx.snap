exports[`initial state renders without crashing 1`] = `
<section
  className="collapsible-filter aggregation">
  <div
    className="layout-row">
    <h5
      className="flex-all" />
    <div
      className="flex-fixed toggle">
      <button
        className="a-btn a-btn__link hover"
        onClick={[Function]}>
        Hide
        <span
          className="cf-icon cf-icon-minus-round" />
      </button>
    </div>
  </div>
  <p />
  <ul>
    <li
      className="flex-fixed layout-row">
      <input
        aria-label="foo"
        checked={undefined}
        className="flex-fixed"
        type="checkbox" />
      <span
        className="flex-all bucket-key">
        foo
      </span>
      <span
        className="flex-fixed bucket-count">
        99
      </span>
    </li>
  </ul>
</section>
`;

exports[`show more only shows the first 6 items of large arrays 1`] = `
<section
  className="collapsible-filter aggregation">
  <div
    className="layout-row">
    <h5
      className="flex-all" />
    <div
      className="flex-fixed toggle">
      <button
        className="a-btn a-btn__link hover"
        onClick={[Function]}>
        Hide
        <span
          className="cf-icon cf-icon-minus-round" />
      </button>
    </div>
  </div>
  <p />
  <ul>
    <li
      className="flex-fixed layout-row">
      <input
        aria-label="alpha"
        checked={undefined}
        className="flex-fixed"
        type="checkbox" />
      <span
        className="flex-all bucket-key">
        alpha
      </span>
      <span
        className="flex-fixed bucket-count">
        99
      </span>
    </li>
    <li
      className="flex-fixed layout-row">
      <input
        aria-label="beta"
        checked={undefined}
        className="flex-fixed"
        type="checkbox" />
      <span
        className="flex-all bucket-key">
        beta
      </span>
      <span
        className="flex-fixed bucket-count">
        99
      </span>
    </li>
    <li
      className="flex-fixed layout-row">
      <input
        aria-label="gamma"
        checked={undefined}
        className="flex-fixed"
        type="checkbox" />
      <span
        className="flex-all bucket-key">
        gamma
      </span>
      <span
        className="flex-fixed bucket-count">
        99
      </span>
    </li>
    <li
      className="flex-fixed layout-row">
      <input
        aria-label="delta"
        checked={undefined}
        className="flex-fixed"
        type="checkbox" />
      <span
        className="flex-all bucket-key">
        delta
      </span>
      <span
        className="flex-fixed bucket-count">
        99
      </span>
    </li>
    <li
      className="flex-fixed layout-row">
      <input
        aria-label="epsilon"
        checked={undefined}
        className="flex-fixed"
        type="checkbox" />
      <span
        className="flex-all bucket-key">
        epsilon
      </span>
      <span
        className="flex-fixed bucket-count">
        99
      </span>
    </li>
  </ul>
  <div
    className="flex-fixed">
    <button
      className="a-btn a-btn__link hover more">
      + Show 
      2
       more
    </button>
  </div>
</section>
`;
